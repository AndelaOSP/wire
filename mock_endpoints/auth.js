const auth = (req, res, next) => {
  const token = req.query.token || req.headers['authorization'];
  if (token == 'null' || token == 'false') {
    res.status(401).send({
      error: true,
      message: 'Sorry. You might not be authorized.',
      status: 'error'
    });
  } else {
    return next();
  }
};

module.exports = auth;
