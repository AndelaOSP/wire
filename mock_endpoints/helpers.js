const getUserName = email => {
  let username = {};
  const [firstName, lastName] = email.split('@')[0].split('.');
  const first = firstName[0].toUpperCase() + firstName.substr(1, firstName.length);
  const last = lastName[0].toUpperCase() + lastName.substr(1, lastName.length);
  username = `${first} ${last}`;
  return username;
};

const getRandomImageUrl = () => {
  const randomNum = Math.floor(Math.random() * 9);
  return `https://randomuser.me/api/portraits/lego/${randomNum}.jpg`;
};

const matchRoleIdToName = (roles, roleId) => {
  let role = roles.find(role => {
    return role.id === roleId;
  });
  return role.name;
};

const matchLocationIdToLocation = (locations, locationId) => {
  let location = locations.find(location => {
    return location.id === locationId;
  });
  let userLocation = Object.assign({}, location);
  delete userLocation.id;
  return userLocation;
};

module.exports = { getUserName, getRandomImageUrl, matchRoleIdToName, matchLocationIdToLocation };
