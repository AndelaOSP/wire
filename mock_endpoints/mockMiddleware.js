const {
  getIncidents,
  getIncident,
  searchIncidents,
  getIncidentNotes,
  getIncidentChats,
  addNote,
  addChat,
  editNote,
  editChat,
  archiveNote,
  archiveChat,
  changeAssignee,
  handleCCd,
  changeStatus,
  getStaff,
  getRoles,
  getLocations,
  addUser,
  editUser,
  searchUser,
  deleteUser,
  login
} = require('./mockControllers');

module.exports = {
  fetchIncidents: (req, res) => {
    setTimeout(() => {
      let query = req.query.q ? req.query.q : '';
      res.send({ data: { incidents: getIncidents(query) }, status: 'success' });
    }, 2000);
  },
  handleSearch: (req, res) => {
    setTimeout(() => {
      let query = req.query.q ? req.query.q : '';
      res.send({ data: { incidents: searchIncidents(query) }, status: 'success' });
    }, 2000);
  },
  fetchIncident: (req, res) => {
    setTimeout(() => {
      let incidentId = parseInt(req.params.id);
      if(getIncident(incidentId) === 'Incident Not Found') {
        return res.status(404).send({ error: true, message: 'Incident not found', status: 'error' });
      }
      res.send({ data: getIncident(incidentId), status: 'success' });
    }, 2000);
  },
  updateIncident: (req, res) => {
    setTimeout(() => {
      let incidentId = parseInt(req.params.id);
      if (req.body.statusId) {
        res.send({ data: changeStatus(incidentId, req.body.statusId), status: 'success' });
      } else if (req.body.assignee) {
        res.send({
          data: changeAssignee(req.body.assignee.incidentId, req.body.assignee.userId, 'assignee'),
          status: 'success'
        });
      } else if (req.body.ccd) {
        res.send({ data: handleCCd(incidentId, req.body.ccd), status: 'success' });
      }
    }, 2000);
  },
  addNoteToIncident: (req, res) => {
    setTimeout(() => {
      let incidentId = parseInt(req.params.id);
      let { userEmail, note } = req.body;
      res.send({ data: addNote(incidentId, userEmail, note), status: 'success' });
    }, 2000);
  },
  fetchIncidentNotes: (req, res) => {
    setTimeout(() => {
      res.send({ data: { notes: getIncidentNotes(parseInt(req.params.id)) }, status: 'success' });
    }, 2000);
  },
  editIncidentNote: (req, res) => {
    setTimeout(() => {
      let noteId = parseInt(req.params.id);
      res.send({ data: editNote(noteId, req.body.note), status: 'success' });
    }, 2000);
  },
  archiveIncidentNote: (req, res) => {
    setTimeout(() => {
      res.send({ data: archiveNote(parseInt(req.params.id)), status: 'success' });
    }, 2000);
  },
  addChatToIncident: (req, res) => {
    setTimeout(() => {
      let incidentId = parseInt(req.params.id);
      let { userId, chat } = req.body;
      res.send({ data: addChat(incidentId, userId, chat), status: 'success' });
    }, 2000);
  },
  fetchIncidentChats: (req, res) => {
    setTimeout(() => {
      res.send({ data: { chats: getIncidentChats(parseInt(req.params.id)) }, status: 'success' });
    }, 2000);
  },
  editIncidentChat: (req, res) => {
    setTimeout(() => {
      let chatId = parseInt(req.params.id);
      res.send({ data: editChat(chatId, req.body.chat), status: 'success' });
    }, 2000);
  },
  archiveIncidentChat: (req, res) => {
    setTimeout(() => {
      res.send({ data: archiveChat(parseInt(req.params.id)), status: 'success' });
    }, 2000);
  },
  fetchStaff: (req, res) => {
    setTimeout(() => {
      res.send({ data: { users: getStaff() }, status: 'success' });
    }, 2000);
  },
  fetchRoles: (req, res) => {
    setTimeout(() => {
      res.send({ data: { roles: getRoles() }, status: 'success' });
    }, 2000);
  },
  fetchLocations: (req, res) => {
    setTimeout(() => {
      res.send({ data: { locations: getLocations() }, status: 'success' });
    }, 2000);
  },
  addUser: (req, res) => {
    setTimeout(() => {
      let { email, roleId, locationId } = req.body;
      res.send({ data: addUser(email, roleId, locationId), status: 'success' });
    }, 2000);
  },
  searchUser: (req, res) => {
    setTimeout(() => {
      let query = req.query.q ? req.query.q : '';
      res.send({ data: { users: searchUser(query) }, status: 'success' });
    }, 2000);
  },
  editUser: (req, res) => {
    setTimeout(() => {
      let userId = parseInt(req.params.id);
      let { roleId } = req.body;
      res.send({ data: editUser(userId, roleId), status: 'success' });
    }, 2000);
  },
  deleteUser: (req, res) => {
    setTimeout(() => {
      let userId = parseInt(req.params.id);
      res.send({ data: deleteUser(userId), status: 'success' });
    }, 2000);
  },
  login: (req, res) => {
    setTimeout(() => {
      if(!login(req.body.email)) {
        return res.status(401).send({ error: true, message: 'User does not exist', status: 'error' });
      }
      res.send({userToken: login(req.body.email), status: 'success' });
    }, 2000);
  }
};
