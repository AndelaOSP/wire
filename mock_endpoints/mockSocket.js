/* eslint-disable no-console */
/**
 * A class for managing server-side socket connections.
 * */
class Socket {
  constructor(server) {
    // Connect socket to running server
    // eslint-disable-next-line global-require
    this.io = require('socket.io')(server);

    // Add cors for socket
    this.io.origins('*:*');

    // Client associations { userId: [socketId, token], ... }
    // REVIEW: Should this be stored in the Db?
    this.clients = {};

    // Set up connections
    this.establishConnection();
  }

  /**
   * Sets up what happens when there is a new connection or disconnection.
   * */
  establishConnection() {
    // Using promise to get the socket value
    this.connection = new Promise((resolve) => {
      // The following happens when a new connection is established.
      this.io.on('connection', (socket) => {
        // Handle new connections
        this.handleNewConnection(socket);

        // Handle disconnections
        this.handleDisconnection(socket);

        // Make socket the Promise data
        resolve(socket);
      });
    });
  }

  /**
   * Handles new client connections
   * @param socket object
   * */
  handleNewConnection(socket) {
    console.log(`[server] A client (${socket.id}) just connected!`);
    // Do something with the data recieved.
    socket.on('new-connection', ({ token }) => {
      // If token is provided
      if (token) {
        // Authenticate user and register to list of connected clients.
        const userDetails = this.getUserDetails(token, socket);

        // Register client
        this.registerClient(userDetails);
      }
    });
  }

  /**
   * Registers client details in this.clients
   * @param userDetails object
   * */
  registerClient(userDetails) {
    userDetails.then(({ socket, user, token }) => {
      this.clients[user.id] = [socket.id, token];
    });
  }


  /**
   * Authenticates a new connected client
   * @param token string
   * @param socket object
   * */
  // eslint-disable-next-line class-methods-use-this
  authenticate(token, socket) {
    return new Promise(resolve => resolve({ socket, user: { id: 1 }, token }));
  }

  /**
   * Does cleanup when a client disconnects
   * @param socket object
   * */
  handleDisconnection(socket) {
    socket.on('disconnect', () => console.log(`[server] A client (${socket.id}) just disconnected!`));

    // Remove client from list of connected clients.
    const entries = Object.entries(this.clients).filter(
      ([, value]) => value[0] !== socket.id,
    );

    this.clients = entries.reduce((prev, [key, value]) => {
      // eslint-disable-next-line no-param-reassign
      prev[key] = value;
      return prev;
    }, {});
  }

  /**
   * Sends a response back to the client to update its UI on notify-cc
   * @param userIds array
   * @param message string
   * */
  notifyCCChange(userIds, message) {
    this.connection.then(() => {
      // Notify each user affected
      this.messageToUsers(userIds, 'notify-cc', message);
    });
  }

  /**
   * Sends a response back to the client to update its UI on notify-cc
   * @param userIds array
   * @param message string
   * */
  messageToUsers(userIds, eventName, message) {
    userIds.forEach((id) => {
      const client = this.clients[id];
      if (client) {
        this.io.to(`${client[0]}`).emit(eventName, { message });
      }
    });
  }
}

module.exports = server => new Socket(server);
