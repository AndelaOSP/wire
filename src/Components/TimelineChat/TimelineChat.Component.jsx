import React, { Component } from 'react';
import { List, ListItem } from 'material-ui/List';
import Avatar from 'material-ui/Avatar';
import TextField from 'material-ui/TextField';
import PropTypes from 'prop-types';
import moment from 'moment';

// styling
import './TimelineChat.scss';

export default class TimelineChat extends Component {
  constructor(props) {
    super(props);
    this.state = {
      message: ''
    };
  }

  handleChange = e => {
    e.preventDefault();
    this.setState({ message: e.target.value });
  };

  handlePostMessage = e => {
    e.preventDefault();
    this.props.sendMessage(this.props.incident.id, this.refs.messageInput.getValue());
    this.setState({ message: '' });
  };

  handleDateString = date => {
    return moment(date).format('[at] h:mm a');
  };

  render() {
    const { chats } = this.props.incident;
    return (
      <div className="chat-container">
        <List className="chat-list">
        {
          chats.length > 0 ? (
            chats.map((chat, i) => {
              return <div className="message-header" key={i}>
              <span>
                  {chat.User ? `${chat.User.username}${this.handleDateString(chat.createdAt)}` : 'You'
                  }
              </span>
                {/* <img src={chat.User.imageUrl} /> */}
                  <div className="message-bubble">
                    <p>{chat.chat}</p>
                  </div>
                </div>;
            })
          ):
          <div>
            <p>No Messages</p>
            </div>
        }
        </List>
        <div className="message-container">
          <img src="/assets/images/clip.svg" color="red" className="notification-icon" />
        <div className="message-input">
          <form onSubmit={this.handlePostMessage}>
            <TextField
              hintText="Type a message"
              value={this.state.message}
              onChange={this.handleChange}
              ref="messageInput"
              underlineShow={false}
              className="text-input"
            />
          </form>
        </div>
          <div className="message-icon">
            <img src="/assets/images/smile.svg" className="message-icon" />
        </div>
         </div>
      </div>
    );
  }
}

TimelineChat.propTypes = {
  incident: PropTypes.object.isRequired,
  sendMessage: PropTypes.func.isRequired
};
