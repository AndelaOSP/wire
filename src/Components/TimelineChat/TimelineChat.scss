@import "../../theme";
.chat-container {
  width: 100%;
  height: 80vh !important;
  overflow: auto;
  .chat-list {
    width: 90%;
    margin: 2rem auto;
  }
  .message-container {
    top: auto !important;
    bottom: 0 !important;
    position: absolute !important;
    .notification-icon {
      cursor: pointer;
      margin-top: 1rem;
      width: 30.4px;
      height: 32.1px;
      margin-left: 1rem;
      position: absolute !important;
    }
    .message-input {
      margin-left: 5rem;
      width: 60vw;
      height: 60px;
      border-radius: 9px;
      background-color: #F1F1F4;
      .text-input {
        border-top-color: map_get($grey, 700) !important;
        width: 100% !important;
        text-align: center;
        padding: .3rem 0 0 1.5rem;
        align-items: center;
      }
    }
    .message-icon {
      cursor: pointer; // padding-right: 1rem;
      width: 25.3px;
      height: 25.3px;
      position: absolute;
      top: 10px;
      right: 13px;
      font-size: 25px;
      cursor: pointer;
      opacity: 0.6;
      padding: .3rem 0 0 1.5rem;
      -webkit-transition: all ease-out 0.3s 0s;
      -o-transition: all ease-out 0.3s 0s;
      transition: all ease-out 0.3s 0s;
      color: #dbdbdb;
    }
  }
}