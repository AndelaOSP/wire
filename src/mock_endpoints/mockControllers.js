const fuzzysearch = require('fuzzysearch');

let { incidents, chats, notes, users, statuses, levels, locations } = require('./mockData');

module.exports = {
  getIncident: incidentId => {
    let incident = incidents.find(incident => {
      return incident.id === incidentId;
    });
    incident['Assignee'] =
      users.find(user => {
        return user.id === incident.assigneeId;
      }) || null;
    incident['User'] =
      users.find(user => {
        return user.id === incident.userId;
      }) || null;
    incident['Status'] =
      statuses.find(status => {
        return status.id === incident.statusId;
      }) || null;
    incident['Level'] =
      levels.find(level => {
        return level.id === incident.levelId;
      }) || null;
    incident['Location'] = locations.find(location => {
      return location.id === incident.locationId;
    });
    return incident;
  },
  getIncidents: query => {
    return incidents.filter(incident => fuzzysearch(query, incident.subject.toLowerCase()) === true).map(incident => {
      incident['Assignee'] =
        users.find(user => {
          return user.id === incident.assigneeId;
        }) || null;
      incident['User'] =
        users.find(user => {
          return user.id === incident.userId;
        }) || null;
      incident['Status'] =
        statuses.find(status => {
          return status.id === incident.statusId;
        }) || null;
      incident['Level'] =
        levels.find(level => {
          return level.id === incident.levelId;
        }) || null;
      return incident;
    });
  },
  getIncidentNotes: incidentId => {
    let incidentNotes = notes.filter(note => {
      return note.incidentId === incidentId;
    });
    incidentNotes.map(
      note =>
        (note['User'] = users.find(user => {
          return user.id === note.userId;
        }))
    );
    return incidentNotes;
  },
  getIncidentChats: incidentId => {
    let incidentChats = chats.filter(chat => {
      return chat.incidentId === incidentId;
    });
    incidentChats.map(chat => {
      chat['User'] = users.find(user => {
        return user.id === chat.userId;
      });
    });
    return incidentChats;
  },
  changeStatus: (incidentId, statusId) => {
    let incident = incidents.find(incident => {
      return incident.id === parseInt(incidentId);
    });
    incident.statusId = statusId;
    incident['Assignee'] =
      users.find(user => {
        return user.id === incident.assigneeId;
      }) || null;
    incident['User'] =
      users.find(user => {
        return user.id === incident.userId;
      }) || null;
    incident['Status'] =
      statuses.find(status => {
        return status.id === statusId;
      }) || null;
    incident['Level'] =
      levels.find(level => {
        return level.id === incident.levelId;
      }) || null;
    incident['Location'] = locations.find(location => {
      return location.id === incident.locationId;
    });
    return incident;
  },
  changeAssignee: (incidentId, assigneeId) => {
    let incident = incidents.find(incident => {
      return incident.id === incidentId;
    });
    incident.assigneeId = assigneeId;
    incident['Assignee'] =
      users.find(user => {
        return user.id === assigneeId;
      }) || null;
    incident['User'] =
      users.find(user => {
        return user.id === incident.userId;
      }) || null;
    incident['Status'] =
      statuses.find(status => {
        return status.id === incident.statusId;
      }) || null;
    incident['Level'] =
      levels.find(level => {
        return level.id === incident.levelId;
      }) || null;
    incident['Location'] = locations.find(location => {
      return location.id === incident.locationId;
    });
    return incident;
  },
  addNote: (incidentId, userId, note) => {
    let newNote = {
      id: notes.length++,
      incidentId,
      userId,
      note
    };
    notes.push(newNote);
    newNote['User'] =
      users.find(user => {
        return user.id === newNote.userId;
      }) || {};
    return newNote;
  },
  editNote: (noteId, note) => {
    let noteToEdit = notes.find(note => {
      return note.id === noteId;
    });
    noteToEdit.note = note;
    noteToEdit['User'] = users.find(user => {
      return user.id === noteToEdit.userId;
    });
    return noteToEdit;
  },
  archiveNote: noteId => {
    return notes.filter(note => {
      return note.id !== noteId;
    });
  },
  getChat: chatId => {
    return chats.find(chat => {
      return chat.id === chatId;
    });
  },
  addChat: (incidentId, userId, chat) => {
    let newChat = {
      id: chats.length++,
      incidentId,
      userId,
      chat
    };
    chats.push(newChat);
    newChat['User'] =
      users.find(user => {
        return user.id === newChat.userId;
      }) || {};
    return newChat;
  },
  editChat: (chatId, chat) => {
    let chatToEdit = chats.find(chat => {
      return chat.id === chatId;
    });
    chatToEdit.chat = chat;
    chatToEdit['User'] = users.find(user => {
      return user.id === chatToEdit.userId;
    });
    return chatToEdit;
  },
  archiveChat: chatId => {
    return chats.filter(chat => {
      return chat.id !== chatId;
    });
  },
  getStaff: () => {
    return users.filter(user => {
      return user.roleId !== 1;
    });
  }
};
