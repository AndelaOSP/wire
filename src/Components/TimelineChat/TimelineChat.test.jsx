import React from 'react';
import { shallow } from 'enzyme';
import shallowToJSON from 'enzyme-to-json';

import TimelineChat from './TimelineChat.Component';
import { testIncidents } from '../../../mock_endpoints/mockData';

describe('Timeline Chat component', () => {
  it('should have all the Timeline Chat content', () => {
    const timelineChat = shallow(
      <TimelineChat incident={testIncidents[0]} sendMessage={() => {}} />,
    );
    const tree = shallowToJSON(timelineChat);
    expect(tree.type).toEqual('div');
    expect(tree.props.className).toEqual('chat-container');
  });
});
