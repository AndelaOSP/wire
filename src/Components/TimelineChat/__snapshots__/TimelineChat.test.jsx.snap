// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline Chat component should have all the Timeline Chat content 1`] = `undefined`;
